compiler/rustc_data_structures/src/fingerprint.rs

```diff
@@ -76,6 +77,33 @@ impl ::std::fmt::Display for Fingerprint {
     }
 }
 
+impl Hash for Fingerprint {
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write_fingerprint(self);
+    }
+}
+
+trait FingerprintHasher {
+    fn write_fingerprint(&mut self, fingerprint: &Fingerprint);
+}
+
+impl<H: Hasher> FingerprintHasher for H {
+    #[inline]
+    default fn write_fingerprint(&mut self, fingerprint: &Fingerprint) {
+        self.write_u64(fingerprint.0);
+        self.write_u64(fingerprint.1);
```

Comment:
Did you try a panic! here instead to try and track down all the cases where we are hashing fingerprints?

Reply:
I hadn't, but I tried just now and it ran into a `StableHasher` for the parent `DefPathHash`:
https://github.com/rust-lang/rust/blob/130359cb05246fcacdde61baa2613419ef6570c7/compiler/rustc_hir/src/definitions.rs#L111-L117

I suppose we could have that hash without the parent at first, and then `Fingerprint::combine` them for the final value. I'll give that a shot and see if anything else comes up.

Reply:
The next one that comes up is `item_ids_hash.hash_stable(...)`:
https://github.com/rust-lang/rust/blob/130359cb05246fcacdde61baa2613419ef6570c7/compiler/rustc_middle/src/ich/impls_hir.rs#L52-L65
I wonder if we could instead add a combine operation directly in the `StableHasher`?

Reply:
Alternatively, we could specialize to allow `Fingerprint`s to be used with `StableHasher` too, and unimplement the rest.

Reply:
Hm, supporting `combine` directly with StableHasher seems like a good idea -- we presumably want that (or similar) a lot.

I'm not sure if combine is "as good" as hashing though, from a "hash quality" perspective. I would sort of assume no because then we wouldn't get any wins from using it...

Reply:
Here's another hit: 
https://github.com/rust-lang/rust/blob/130359cb05246fcacdde61baa2613419ef6570c7/compiler/rustc_query_system/src/dep_graph/graph.rs#L891
... where `DepNode<K>` is a `K` and a `Fingerprint`.

I'm inclined to let all of these hash normally for now. `Unhasher` is already a bit hacky itself...

Reply:
If we wanted though, `DepNode<K>` could ignore its `K` for hashing purposes.

Reply:
It does sound like there's sort of a lot of potential here but the current `Hash/Hasher` API doesn't readily allow specializing like we want to.

I was looking at the Hasher API, and it feels like it might be worth adding something like a `write_hash` or similar, where the Hasher can expect the incoming value to already be "hashed" or at least nicely distributed across the range. To start we could just take a u64 since that's what `Hasher::finish()` returns, though e.g. for Fingerprint we really want u128. Maybe `fn write_hash(impl Into<u128>)` makes sense, not sure.

I am leaning towards saying that we should just merge this PR as-is: it seems like a clear, if small, win, and while there may be more hidden through careful hash-skipping it's probably better to evaluate each in a standalone manner, particularly given the relative complexity of the Unhasher design. We can consider other improvements, like the one I suggested in the previous paragraph, later.

Reply:
> Hm, supporting `combine` directly with StableHasher seems like a good idea -- we presumably want that (or similar) a lot.

FWIW, I was just wanting something like this for span's hash implementation, which currently re-hashes a hash of the file name.